File: PhaseGate/PhoneBook/PhoneBook.java
import java.util.Scanner;

public class PhoneBook {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        
        System.out.println("Enter the number of contacts:");
        int numberOfContacts = input.nextInt();
        input.nextLine();

        String[][] contacts = new String[numberOfContacts][3];

        String menuOption;

        do {
            System.out.println("Select a menu (add, find, delete, exit): ");
            menuOption = input.nextLine();

            switch (menuOption) {
                case "add":
                    addContact(input, contacts);
                    break;
                case "find":
                    findContact(input, contacts);
                    break;
                case "delete":
                    deleteContact(input, contacts);
                    break;
                case "exit":
                    System.out.println("Exiting the phone book app.");
                    break;
                default:
                    System.out.println("Unknown command.");
            }
        } while (!menuOption.equals("exit"));

        input.close();
    }


    public static void addContact(Scanner input, String[][] contacts) {
        System.out.println("Enter first name:");
        String fname = input.nextLine();
        System.out.println("Enter last name:");
        String lname = input.nextLine();
        System.out.println("Enter phone number:");
        String phone = input.nextLine();

        boolean added = false;
        for (int i = 0; i < contacts.length; i++) {
            if (contacts[i][0] == null) {
                contacts[i][0] = fname;
                contacts[i][1] = lname;
                contacts[i][2] = phone;
                added = true;
                System.out.printf("Contact added: " + fname + " " + lname);
                break;
            }
        }
        if (!added) {
            System.out.println("Contact list is full.");
        }
    }


    public static void findContact(Scanner input, String[][] contacts) {
        System.out.println("Enter name to find contact:");
        String nameToFind = input.nextLine();
        boolean found = false;

        for (int i = 0; i < contacts.length; i++) {
            if (contacts[i][0] != null && (contacts[i][0].equalsIgnoreCase(nameToFind) || contacts[i][1].equalsIgnoreCase(nameToFind))) {
                System.out.printf("Contact found! Name: %s %s, Phone: %s%n", contacts[i][0], contacts[i][1], contacts[i][2]);
                found = true;
                break;
            }
        }
        if (!found) {
            System.out.println("Contact not found.");
        }
    }


    public static void deleteContact(Scanner input, String[][] contacts) {
        System.out.println("Enter name to delete:");
        String contactName = input.nextLine();
        boolean deleted = false;

        for (int i = 0; i < contacts.length; i++) {
            if (contacts[i][0] != null && (contacts[i][0].equalsIgnoreCase(contactName) || contacts[i][1].equalsIgnoreCase(contactName))) {
                contacts[i][0] = null;
                contacts[i][1] = null;
                contacts[i][2] = null;
                System.out.println("Contact deleted: " + contactName);
                deleted = true;
                break;
            }
        }
        if (!deleted) {
            System.out.println("Contact not found for deletion.");
        }
    }
}

File: PhaseGate/PhoneBook/phone_book.py
contacts = []

def validate_phone_number() -> str:
    while True:
        phone = input("Enter phone number: ")
        if len(phone) != 11 or not phone.isdigit():
            print("Invalid phone number! Please enter an 11-digit numeric phone number.")
        else:
            return phone


def add_contact(contacts: list) -> str:
    first_name = str(input("Enter first name: ")) or None
    last_name = str(input("Enter last name: ")) or None
    phone_number = validate_phone_number()
    contacts.append(dict(
        first_name = first_name,
        last_name = last_name,
        phone_number = phone_number
    ))
    return f"Contact added {first_name} {last_name}"


def search_contact(name: str) -> str:
    for contact in contacts:
        if name.lower() == contact.get("first_name") or name.lower() == contact.get("last_name"):
            return f"Contact found! name: {contact.get('first_name')} {contact.get('last_name')}, phone: {contact.get('phone_number')}"
    return "Contact not found."


def edit_contact(name: str) -> str:
    for contact in contacts:
        if name.lower() == contact.get("first_name").lower() or name.lower() == contact.get("last_name").lower():
            new_number = validate_phone_number()
            contact["phone_number"] = new_number
            return f"Contact updated! Name: {contact.get('first_name')} {contact.get('last_name')}, Phone: {contact.get('phone_number')}"
    return "Contact not found to edit."


def delete_contact(name: str) -> str:
    for index, contact in enumerate(contacts):
        if name.lower() == contact.get("first_name").lower() or name.lower() == contact.get("last_name").lower():
            full_name = f"{contact.get('first_name')} {contact.get('last_name')}"
            option = input(f"Warning! Are you sure you wish to delete {full_name}? (y/n): ")
            if option.lower() == "y" or option.lower() == "yes":
                del contacts[index]
                return f"Contact {full_name} deleted successfully."
            else:
                return f"Delete for {full_name} was canceled."
    return "Contact not found to delete."


while True:
    menu = str(input("Select from the menu (add, search, edit, delete, exit): "))
    
    match menu.lower():
        case "add":
            print(add_contact(contacts))
        case "search":
            find_name = str(input("Enter first/last name to search contact: "))
            print(search_contact(find_name))
        case "edit":
            find_name = str(input("Enter first/last name to edit contact: "))
            print(edit_contact(find_name))
        case "delete":
            find_name = str(input("Enter first/last name to delete contact: "))
            print(delete_contact(find_name))
        case _:
            print("Invalid option")
    if menu == "exit":
        break